src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -23,502 +23,309 @@ namespace Roslyn.Utilities
+        public void WriteDouble(double value) => _writer.Write(value);
+        public void WriteSingle(float value) => _writer.Write(value);
+        public void WriteInt32(int value) => _writer.Write(value);
+        public void WriteInt64(long value) => _writer.Write(value);
+        public void WriteSByte(sbyte value) => _writer.Write(value);
+        public void WriteInt16(short value) => _writer.Write(value);
+        public void WriteUInt32(uint value) => _writer.Write(value);
+        public void WriteUInt64(ulong value) => _writer.Write(value);
+        public void WriteUInt16(ushort value) => _writer.Write(value);
+        public void WriteString(string value) => WriteStringValue(value);
+
+        public void WriteValue(object value)
         {
-            _writer.Write(value);
-        }
-
-        public override void WriteUInt32(uint value)
-        {
-            _writer.Write(value);
-        }
+            Debug.Assert(value == null || !value.GetType().GetTypeInfo().IsEnum, "Enum should not be written with WriteValue.  Write them as ints instead.");
 
-        public override void WriteUInt64(ulong value)
-        {
-            _writer.Write(value);
-        }
+            if (value == null)
+            {
+                _writer.Write((byte)EncodingKind.Null);
+                return;
+            }
 
-        public override void WriteUInt16(ushort value)
-        {
-            _writer.Write(value);
-        }
+            var type = value.GetType();
+            var typeInfo = type.GetTypeInfo();
+            Debug.Assert(!typeInfo.IsEnum, "Enums should not be written with WriteObject.  Write them out as integers instead.");
 
-        public override void WriteString(string value)
-        {
-            WriteStringValue(value);
-        }
+            // Perf: Note that JIT optimizes each expression value.GetType() == typeof(T) to a single register comparison.
+            // Also the checks are sorted by commonality of the checked types.
 
-        public override void WriteValue(object value)
-        {
-            if (_recursive)
-            {
-                WriteVariant(Variant.FromBoxedObject(value));
-            }
-            else
+            // The primitive types are Boolean, Byte, SByte, Int16, UInt16, Int32, UInt32, Int64, UInt64, IntPtr, UIntPtr, Char, Double, and Single.
+            if (typeInfo.IsPrimitive)
             {
-                _valueStack.Push(Variant.FromBoxedObject(value));
-                Emit();
-            }
-        }
+                // Note: int, double, bool, char, have been chosen to go first as they're they
+                // common values of literals in code, and so would be hte likely hits if we do
+                // have a primitive type we're serializing out.
+                if (value.GetType() == typeof(int))
+                {
+                    WriteEncodedInt32((int)value);
+                    return;
```

Comment:
because i don't want them to fall out the bottom.  Look at the whole section.  Once we write, we have to stop immediately.   If we had if/elses eventually we'd fall out and then excute a bunch more code.